File: padam_padam/HW3/40_reverse_linked_list.py
'''40. reverse linked list'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        node = head
        pre = None

        while node:
            next = node.next
            node.next = pre

            pre = node
            node = next
        
        return pre

# actually recusrive function seems easier for me..
def reverseList(self, head) -> ListNode:
    def reversef(node, prev=None):
        if not node:
            return prev
        next, node.next = node.next, prev
        return reversef(next, node)

    return reversef(head)

File: juyomo/hw3/40_reverse_linked_list.cpp
// Author: Juyoung Moon
// https://leetcode.com/problems/reverse-linked-list/

// HOCO Mentoring HW 3
// https://github.com/juyomo/hoco-mentoring

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) {
            return head;
        }

        ListNode* curr = head;
        ListNode* prev = nullptr;
        ListNode* next = head->next;

        while (curr != nullptr) {
            curr->next = prev;
            
            prev = curr;
            curr = next;
            if (curr != nullptr) {
                next = curr->next;
            }
        }
        
        return prev;
    }
};  
